src/Components/Server/src/Builder/ServerRazorComponentsEndpointConventionBuilderExtensions.cs

```diff
@@ -17,7 +20,38 @@ public static class ServerRazorComponentsEndpointConventionBuilderExtensions
     /// <returns>The <see cref="RazorComponentsEndpointConventionBuilder"/>.</returns>
     public static RazorComponentsEndpointConventionBuilder AddInteractiveServerRenderMode(this RazorComponentsEndpointConventionBuilder builder)
     {
-        ComponentEndpointConventionBuilderHelper.AddRenderMode(builder, new InternalServerRenderMode());
+        return AddInteractiveServerRenderMode(builder, null);
+    }
+
+    /// <summary>
+    ///Maps the Blazor <see cref="Hub" /> to the default path.
+    /// </summary>
+    /// <param name="builder">The <see cref="RazorComponentsEndpointConventionBuilder"/>.</param>
+    /// <param name="callback">A callback to configure server endpoint options.</param>
+    /// <returns>The <see cref="ComponentEndpointConventionBuilder"/>.</returns>
+    public static RazorComponentsEndpointConventionBuilder AddInteractiveServerRenderMode(
+        this RazorComponentsEndpointConventionBuilder builder,
+        Action<ServerComponentsEndpointOptions>? callback = null)
+    {
+        var options = new ServerComponentsEndpointOptions();
+        callback?.Invoke(options);
+
+        ComponentEndpointConventionBuilderHelper.AddRenderMode(builder, new InternalServerRenderMode(options));
+
+        if ((options.EnableWebSocketCompression || options.ConfigureConnectionOptions is not null) && options.ContentSecurityFrameAncestorPolicy != null)
+        {
+            builder.AddEndpointFilter(new RequireCspFilter(options.ContentSecurityFrameAncestorPolicy));
+        }
+
         return builder;
     }
+
+    private class RequireCspFilter(string policy) : IEndpointFilter
+    {
+        public ValueTask<object?> InvokeAsync(EndpointFilterInvocationContext context, EndpointFilterDelegate next)
+        {
+            context.HttpContext.Response.Headers.Add("Content-Security-Policy", $"frame-ancestors {policy}");
```

Comment:
Based on [this conversation](https://github.com/w3c/webappsec-csp/issues/215) I get the impression that, if the framework sends this header, it doesn't preclude the developer from independently setting any other `Content-Security-Policy` header(s), and the browser will enforce all of them, meaning that at most, we are just adding restrictions to any restrictions that are set elsewhere.

If that matches your understanding, that sounds fine to me. As long as us setting this value by default won't be confusing or breaking for people who already set some other CSP header, this looks good to me.

Reply:
Yes, they accumulate so it's common to break the policy into multiple headers so as to deal with different directives independently. They all get enforced, which means that the strongest directive always wins. (meaning that you can't put a `frame-ancestors` in a separate policy that weakens it and that if you have other policies to control scripts, styles, etc. This won't get in the way.